src/AppInstallerCommonCore/Public/winget/WindowsFeature.h

```diff
@@ -194,5 +144,46 @@ namespace AppInstaller::WindowsFeature
         void OpenSession();
         void CloseSession();
         void Shutdown();
+
+        template<typename FuncType>
+        FuncType GetProcAddressHelper(HMODULE module, LPCSTR functionName);
+    };
+
+    /// <summary>
+    /// Struct representation of a single Windows Feature.
+    /// </summary>
+    struct WindowsFeature
+    {
+        // This default constructor is only used for mocking unit tests.
+        WindowsFeature() = default;
+        WindowsFeature(std::shared_ptr<DismHelper> dismHelper, const std::string& name);

```

Comment:
I left the constructor as is to be able to get a shared_ptr to DismHelper. I changed DismHelper to derive from `enable_shared_from_this` so that we can safely get another shared_ptr as suggested.

Reply:
The awkward part is the call site; the need to pass the `shared_ptr` yourself.  (I should have put the comment there I suppose...)  I wasn't suggesting that you should take the constructor away (although you could make it protected and friend the other way).  I *was* suggesting that you add a function like:
```
WindowsFeature DismHelper::GetFeature(std::string featureName);
```
So that the call site was less awkward, like:
```
WindowsFeature::WindowsFeature windowsFeature = dismHelper->GetFeature(featureName};
```
If you did make the constructor protected, you then don't have to worry about the `shared_ptr` being null.

Reply:
I agree the call site is weird passing shared ptrs around. I feel that this is important if dismhelper were to be extended to other scenarios so I changed it based on your suggestions. I created a helper function within dismhelper that creates the windows object for you, calling the constructor with the shared ptr. I made the WindowsFeature constructor protected and made dismHelper a friend so that only the dismhelper can create the WindowsFeature object.